astroid/bases.py

```diff
@@ -129,12 +129,15 @@ def _infer_stmts(
 ) -> collections.abc.Generator[InferenceResult, None, None]:
     """Return an iterator on statements inferred by each statement in *stmts*."""
     inferred = False
+    constraint_failed = False
     if context is not None:
         name = context.lookupname
         context = context.clone()
+        constraints = context.constraints.get(name, {})
     else:
         name = None
         context = InferenceContext()
+        constraints = {}
```

Comment:
I like to initialize container types unconditionally. That opens the possibility of adding a type hint which will help type checkers infer other variables later. Like `stmt_constraints` with is currently `set[Unknown]`. That should get better with
```py
constraints: dict[str, dict[nodes.If, constraints.Constraint]] = {}
```